File: iOS8_Adv_Training/Sprite/Sprite/GameScene.m
//
//  GameScene.m
//  Sprite
//
//  Created by 新居雅行 on 2014/10/30.
//  Copyright (c) 2014年 msyk.net. All rights reserved.
//

#import "GameScene.h"

@interface GameScene () <SKPhysicsContactDelegate>

@property SKLabelNode *myLabel;
@property SKSpriteNode *ship;
@property NSTimer *timer;
@property long count;

- (void)flyingBox;
- (void)repeatTask: (NSTimer *)timer;

@end

@implementation GameScene

- (CGFloat)positionControl
{
    return 0.0;
}

- (void)setPositionControl:(CGFloat) value
{
    CGFloat newX = CGRectGetMidX(self.frame)
    + CGRectGetWidth(self.frame) * (value - 0.5);
    CGFloat newY = self.ship.position.y;
    self.ship.position = CGPointMake(newX, newY);
}

- (void)didBeginContact:(SKPhysicsContact *)contact
{
    [self.timer invalidate];
    self.myLabel.text = @"Game Over!";
    self.backgroundColor = [UIColor redColor];
}

-(void)didMoveToView:(SKView *)view {
    
    srand([NSDate date].timeIntervalSince1970);
    self.physicsWorld.gravity = CGVectorMake(0,0);
    self.physicsWorld.contactDelegate = self;

    self.myLabel = [SKLabelNode labelNodeWithFontNamed:@"Chalkduster"];
    
    self.myLabel.text = @"Hello, World!";
    self.myLabel.fontSize = 65;
    self.myLabel.position = CGPointMake(CGRectGetMidX(self.frame),
                                        CGRectGetMidY(self.frame));
    
    [self addChild: self.myLabel];
    
    self.ship = [SKSpriteNode spriteNodeWithImageNamed:@"Spaceship"];
    
    self.ship.xScale = 0.3;
    self.ship.yScale = 0.3;
    self.ship.position = CGPointMake(CGRectGetMidX(self.frame),
                                     CGRectGetMidY(self.frame) - 150);
    self.ship.physicsBody = [SKPhysicsBody bodyWithRectangleOfSize: self.ship.size];
    self.ship.physicsBody.categoryBitMask = 0x01;
    self.ship.physicsBody.collisionBitMask = 0x03;
    self.ship.physicsBody.contactTestBitMask = 0x02;

    [self addChild: self.ship];
    
//    [self flyingBox];
    self.timer = [NSTimer scheduledTimerWithTimeInterval: 1.0
                                                  target: self
                                                selector: @selector(repeatTask:)
                                                userInfo: nil
                                                 repeats: YES];
    
}

- (void)repeatTask: (NSTimer *)timer
{
    [self flyingBox];
    self.count++;
}

-(void)update:(CFTimeInterval)currentTime {
    /* Called before each frame is rendered */
}

- (void)flyingBox
{
    CGFloat range = 1000;
    CGFloat minX = CGRectGetMinX(self.frame);
//    CGFloat minY = CGRectGetMinY(self.frame);
//    CGFloat midX = CGRectGetMidX(self.frame);
    CGFloat midY = CGRectGetMidY(self.frame);
    CGFloat maxX = CGRectGetMaxX(self.frame);
//    CGFloat maxY = CGRectGetMaxY(self.frame);
    CGFloat xValue = (self.count % 2 == 0) ? maxX : minX;
    CGFloat xValueMove = (self.count % 2 == 0) ? minX : maxX;
//    CGFloat xValue = maxX;
//    CGFloat xValueMove = minX;
    CGFloat yValue = midY + 1000 + range * rand() / RAND_MAX;
    CGFloat yValueMove = midY - 1000 - range * rand() / RAND_MAX;

    UIColor *boxColor = [UIColor blueColor];
    CGSize boxSize = CGSizeMake(30.0, 30.0);
    CGPoint boxPos = CGPointMake(xValue, yValue);
    SKSpriteNode *box = [SKSpriteNode spriteNodeWithColor: boxColor
                                                     size: boxSize];
    box.position = boxPos;
    
    SKAction *rotateAction = [SKAction rotateByAngle: M_PI/2.0
                                            duration: 0.5];
    [box runAction: [SKAction repeatActionForever: rotateAction]];
    
    CGPoint movePos = CGPointMake(xValueMove, yValueMove);
    SKAction *moveAction = [SKAction moveTo: movePos
                                   duration: 6.0];
    [box runAction: moveAction completion: ^(){
        [box removeFromParent];
    }];
    [box runAction: moveAction];
    
    box.physicsBody = [SKPhysicsBody bodyWithRectangleOfSize: box.size];
    box.physicsBody.categoryBitMask = 0x02;
//    box.physicsBody.collisionBitMask = 0x01;
//    box.physicsBody.contactTestBitMask = 0x01;
    
    [self addChild: box];
}

@end

File: iOS8_Adv_Training/GameTest/GameTest/GameScene.swift
//
//  GameScene.swift
//  GameTest
//
//  Created by 新居雅行 on 2014/10/26.
//  Copyright (c) 2014年 msyk.net. All rights reserved.
//

import SpriteKit

class GameScene: SKScene, SKPhysicsContactDelegate {
    
    var message: SKLabelNode?
    var timer: NSTimer?
    var count: Int32 = 0
    var ship: SKSpriteNode = SKSpriteNode(imageNamed:"Spaceship")
    var positionControl: CGFloat {
        set(aValue)    {
            let newX = self.frame.midX + self.frame.width * (aValue - 0.5)
            let newY = ship.position.y
            ship.position = CGPointMake(newX, newY)
        }
        get    {
            return 0.0
        }
    }
    
    override func didMoveToView(view: SKView) {
        /* Setup your scene here */
        
        self.physicsWorld.gravity = CGVectorMake(0,0);
        self.physicsWorld.contactDelegate = self;
        
        message = SKLabelNode(fontNamed:"Chalkduster")
        if let message = message{
        message.text = "Go ahead!";
        message.fontSize = 65;
        message.position = CGPoint(x: self.frame.midX, y: self.frame.midY);
            self.addChild(message)
        }
        
        
        spaceShip()
        
        timer = NSTimer.scheduledTimerWithTimeInterval(0.2, target: self, selector: "repeatingTask:", userInfo: nil, repeats: true)
        println(self.frame)
        println(self.anchorPoint)
        println(self.size)
    }
    
    func didBeginContact(contact: SKPhysicsContact)    {
        println("#####")
        timer?.invalidate()
        message?.text = "Game Over!";
        self.backgroundColor = UIColor.redColor()
    }
    
    func repeatingTask(timer: NSTimer)  {
        flyingBall()
    }
    
    func randFloat() -> CGFloat    {
        return CGFloat(rand()) / CGFloat(RAND_MAX)
    }
    
    func flyingBall()   {
        let newBall = SKSpriteNode(color: UIColor.blueColor(), size: CGSizeMake(30, 30))
        let range: CGFloat = 1000;
        let middleY = self.frame.midX
        let yValue = middleY + range * randFloat()
        let yValueMove = middleY - range * randFloat()
        let xValue = (count % 2 == 0) ? self.frame.maxX : self.frame.minX;
        let xValueMove = (count % 2 == 0) ? self.frame.minX : self.frame.maxX;
        //newBall.position = CGPointMake(CGRectGetMidX(self.frame),CGRectGetMidY(self.frame));
        newBall.position = CGPointMake(xValue, yValue);
        let rotateAction = SKAction.rotateByAngle(CGFloat(M_PI), duration:0.2)
        let moveAction = SKAction.moveTo(CGPointMake(xValueMove,yValueMove), duration: 3.0)
        let repeatAction = SKAction.repeatActionForever(rotateAction)
        newBall.runAction(repeatAction)
        newBall.runAction(moveAction, completion: {
            println("=====")
            newBall.removeFromParent()
        })
        newBall.physicsBody = SKPhysicsBody(rectangleOfSize: newBall.size);
            newBall.physicsBody?.categoryBitMask = 0x02;
            newBall.physicsBody?.collisionBitMask = 0x03;
            newBall.physicsBody?.contactTestBitMask = 0x03;
        self.addChild(newBall)
        count++
    }
    
    func spaceShip()    {
       // let ship = SKSpriteNode(imageNamed:"Spaceship")
        ship.xScale = 0.3
        ship.yScale = 0.3
        ship.position = CGPointMake(CGRectGetMidX(self.frame),CGRectGetMidY(self.frame))
        ship.physicsBody = SKPhysicsBody(rectangleOfSize: ship.size);
        ship.physicsBody?.categoryBitMask = 0x01;
        ship.physicsBody?.collisionBitMask = 0x03;
        ship.physicsBody?.contactTestBitMask = 0x03;
        self.addChild(ship)
    }
    
    override func touchesBegan(touches: NSSet, withEvent event: UIEvent) {
        /* Called when a touch begins */
        
        //        for touch: AnyObject in touches {
        //            let location = touch.locationInNode(self)
        //
        //            let sprite = SKSpriteNode(imageNamed:"Spaceship")
        //
        //            sprite.xScale = 0.5
        //            sprite.yScale = 0.5
        //            sprite.position = location
        //
        //            let action = SKAction.rotateByAngle(CGFloat(M_PI), duration:1)
        //
        //            sprite.runAction(SKAction.repeatActionForever(action))
        //
        //            self.addChild(sprite)
        //        }
    }
    
    override func update(currentTime: CFTimeInterval) {
        /* Called before each frame is rendered */
    }
}
